yjit_codegen.c

```diff
@@ -563,102 +570,223 @@ guard_self_is_object(codeblock_t *cb, x86opnd_t self_opnd, uint8_t *side_exit, c
+            RUBY_ASSERT(false && "unimplemented jump kind");
+            break;
+    };
 
+    if (ctx->chain_depth < depth_limit) {
+        ctx_t deeper = *ctx;
+        deeper.chain_depth++;
+
+        gen_branch(
+            ctx,
+            (blockid_t) { jit->iseq, jit->insn_idx },
+            &deeper,
+            BLOCKID_NULL,
+            NULL,
+            target0_gen_fn
+        );
+    }
+    else {
+        target0_gen_fn(cb, side_exit, NULL, SHAPE_DEFAULT);
+    }
+}
 
 
+bool rb_iv_index_tbl_lookup(struct st_table *iv_index_tbl, ID id, struct rb_iv_index_tbl_entry **ent); // vm_insnhelper.c
 
+enum {
+    getivar_max_depth = 10 // up to 5 different classes, and embedded or not for each
+};
 
+static codegen_status_t
+gen_getinstancevariable(jitstate_t* jit, ctx_t* ctx)
+{
+    // Defer compilation so we can peek at the topmost object
+    if (!jit_at_current_insn(jit)) {
+        defer_compilation(jit->block, jit->insn_idx, ctx);
+        return YJIT_END_BLOCK;
+    }
 
+    // Specialize base on the compile time self
+    VALUE self_val = jit_peek_at_self(jit, ctx);
+    VALUE self_klass = rb_class_of(self_val);
 
+    uint8_t *side_exit = yjit_side_exit(jit, ctx);
 
     // If the class uses the default allocator, instances should all be T_OBJECT
     // NOTE: This assumes nobody changes the allocator of the class after allocation.
     //       Eventually, we can encode whether an object is T_OBJECT or not
     //       inside object shapes.
-    if (rb_get_alloc_func(ic->entry->class_value) != rb_class_allocate_instance) {
-        return YJIT_CANT_COMPILE;
+    if (rb_get_alloc_func(self_klass) != rb_class_allocate_instance) {
+        jmp_ptr(cb, side_exit);
+        return YJIT_END_BLOCK;
     }
+    RUBY_ASSERT(BUILTIN_TYPE(self_val) == T_OBJECT); // because we checked the allocator
 
-    uint32_t ivar_index = ic->entry->index;
-
+    ID id = (ID)jit_get_arg(jit, 0);
+    struct rb_iv_index_tbl_entry *ent;
+    struct st_table *iv_index_tbl = ROBJECT_IV_INDEX_TBL(self_val);
     // Create a size-exit to fall back to the interpreter
-    uint8_t* side_exit = yjit_side_exit(jit, ctx);
 
-    // Load self from CFP
-    mov(cb, REG0, member_opnd(REG_CFP, rb_control_frame_t, self));
+    if (iv_index_tbl && rb_iv_index_tbl_lookup(iv_index_tbl, id, &ent)) {
+        uint32_t ivar_index = ent->index;
 
-    guard_self_is_object(cb, REG0, side_exit, ctx);
+        // Load self from CFP
+        mov(cb, REG0, member_opnd(REG_CFP, rb_control_frame_t, self));
 
-    // Bail if receiver class is different from compiled time call cache class
-    x86opnd_t klass_opnd = mem_opnd(64, REG0, offsetof(struct RBasic, klass));
-    mov(cb, REG1, klass_opnd);
-    x86opnd_t serial_opnd = mem_opnd(64, REG1, offsetof(struct RClass, class_serial));
-    cmp(cb, serial_opnd, imm_opnd(ic->entry->class_serial));
-    jne_ptr(cb, side_exit);
+        guard_self_is_object(cb, REG0, COUNTED_EXIT(side_exit, getivar_se_self_not_heap), ctx);
```

Comment:
Is `guard_self_is_object` not needed for the other cases?

Reply:
No. The other case takes the side exit and doesn't read self.